File: packages/typescript/src/router.ts
/**
 * Router for path-based multi-agent routing.
 *
 * Mounts multiple A2A agents under a single Express server at different URL prefixes.
 * Each sub-agent retains its own executor, middleware, auth, and task store.
 *
 * Example:
 *   import { Router } from 'a2a-lite';
 *
 *   const weather = new Agent({ name: "WeatherAgent", description: "Weather forecasts" });
 *   const hotels = new Agent({ name: "HotelAgent", description: "Hotel search" });
 *
 *   weather.skill("forecast", async ({ city }) => `Sunny in ${city}`);
 *   hotels.skill("search", async ({ city }) => [{ name: "Grand Hotel", city }]);
 *
 *   const router = new Router();
 *   router.mount("/weather", weather);
 *   router.mount("/hotels", hotels);
 *   router.run();
 *
 * The merged agent card is at /.well-known/agent.json.
 */

import express, { Express } from 'express';
import type { Agent } from './agent.js';

export class Router {
  private mounts: Array<[string, Agent]> = [];

  /**
   * Mount an agent at a URL prefix.
   *
   * @param prefix - URL prefix (e.g., "/weather"). Leading slash is added if missing.
   * @param agent - The Agent instance to mount.
   */
  mount(prefix: string, agent: Agent): this {
    if (!prefix.startsWith('/')) prefix = '/' + prefix;
    prefix = prefix.replace(/\/$/, '');
    this.mounts.push([prefix, agent]);
    return this;
  }

  /**
   * Build the merged agent card combining skills from all mounted agents.
   */
  buildMergedCard(host = 'localhost', port = 8787): Record<string, unknown> {
    const allSkills: Array<Record<string, unknown>> = [];
    const names: string[] = [];
    const descriptions: string[] = [];
    let hasStreaming = false;

    for (const [prefix, agent] of this.mounts) {
      names.push(agent.name);
      descriptions.push(agent.description);
      const card = agent.buildAgentCard(host, port);
      if (card.capabilities?.streaming) hasStreaming = true;
      for (const skill of card.skills ?? []) {
        allSkills.push({
          id: `${prefix.replace(/^\//, '')}/${skill.id}`,
          name: skill.name,
          description: `[${agent.name}] ${skill.description ?? ''}`,
          tags: skill.tags ?? [],
        });
      }
    }

    return {
      name: names.join(' + ') || 'Router',
      description: descriptions.join('; ') || 'Multi-agent router',
      version: '1.0.0',
      url: `http://${host}:${port}`,
      protocolVersion: '0.3.0',
      capabilities: { streaming: hasStreaming, pushNotifications: false },
      defaultInputModes: ['text'],
      defaultOutputModes: ['text'],
      skills: allSkills,
    };
  }

  /**
   * Build an Express app with all agents mounted at their prefixes.
   */
  buildApp(host = 'localhost', port = 8787): Express {
    const app = express();
    const mergedCard = this.buildMergedCard(host, port);

    // Merged agent card at root
    app.get('/.well-known/agent.json', (_req, res) => {
      res.json(mergedCard);
    });

    // Mount each agent's Express app at its prefix
    for (const [prefix, agent] of this.mounts) {
      app.use(prefix, agent.buildApp());
    }

    return app;
  }

  /**
   * Start the router server.
   */
  async run(options: { host?: string; port?: number; logLevel?: string } = {}): Promise<void> {
    const { host = '0.0.0.0', port = 8787 } = options;
    const displayHost = host === '0.0.0.0' ? 'localhost' : host;
    const app = this.buildApp(displayHost, port);

    console.log(`\nA2A Lite Router → http://${displayHost}:${port}`);
    for (const [prefix, agent] of this.mounts) {
      const skillCount = agent.buildAgentCard().skills?.length ?? 0;
      console.log(`  ${prefix} → ${agent.name} (${skillCount} skills)`);
    }

    await new Promise<void>((resolve) => {
      app.listen(port, host, () => resolve());
    });
  }
}

File: packages/python/src/a2a_lite/router.py
"""
AgentRouter for path-based multi-agent routing.

Mounts multiple agents under a single Starlette app at different URL prefixes.
Each sub-agent retains its own executor, middleware, auth, and task store.

Example:
    from a2a_lite import Agent
    from a2a_lite.router import Router

    weather = Agent(name="WeatherAgent", description="Weather forecasts")
    hotels = Agent(name="HotelAgent", description="Hotel search")

    @weather.skill("forecast")
    async def forecast(city: str) -> str:
        return f"Sunny in {city}"

    @hotels.skill("search")
    async def search(city: str) -> list:
        return [{"name": "Grand Hotel", "city": city}]

    router = Router()
    router.mount("/weather", weather)
    router.mount("/hotels", hotels)
    router.run(port=8787)

The merged agent card is available at /.well-known/agent.json.
Individual agents remain accessible at:
  - /weather/.well-known/agent.json
  - /hotels/.well-known/agent.json
"""

from __future__ import annotations

import uvicorn
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.routing import Mount, Route

from .agent import Agent


class Router:
    """
    Path-based router that mounts multiple A2A agents under a single server.

    Each agent is mounted at a URL prefix. The merged agent card combines
    skills from all sub-agents and is served at /.well-known/agent.json.

    Example:
        router = Router()
        router.mount("/weather", weather_agent)
        router.mount("/hotels", hotel_agent)
        router.run()
    """

    def __init__(self) -> None:
        self._mounts: list[tuple[str, Agent]] = []

    def mount(self, prefix: str, agent: Agent) -> None:
        """
        Mount an agent at a URL prefix.

        Args:
            prefix: URL prefix (e.g., "/weather"). Must start with "/".
            agent: The Agent instance to mount.
        """
        if not prefix.startswith("/"):
            prefix = "/" + prefix
        prefix = prefix.rstrip("/")
        self._mounts.append((prefix, agent))

    def build_merged_card(self, host: str = "localhost", port: int = 8787) -> dict:
        """
        Build a merged agent card combining all mounted agents' skills.

        Each skill ID is prefixed with the mount path to avoid collisions.

        Args:
            host: Hostname for the agent URL.
            port: Port for the agent URL.

        Returns:
            A dict representing the merged agent card.
        """
        all_skills = []
        names = []
        descriptions = []

        for prefix, agent in self._mounts:
            names.append(agent.name)
            descriptions.append(agent.description)
            for skill_def in agent._skills.values():
                all_skills.append(
                    {
                        "id": f"{prefix.lstrip('/')}/{skill_def.name}",
                        "name": skill_def.name,
                        "description": f"[{agent.name}] {skill_def.description}",
                        "tags": skill_def.tags,
                        "inputModes": ["application/json"],
                        "outputModes": ["application/json"],
                    }
                )

        return {
            "name": " + ".join(names) if names else "Router",
            "description": "; ".join(descriptions) if descriptions else "Multi-agent router",
            "version": "1.0.0",
            "url": f"http://{host}:{port}",
            "protocolVersion": "0.3.0",
            "capabilities": {
                "streaming": any(agent._has_streaming for _, agent in self._mounts),
                "pushNotifications": False,
            },
            "defaultInputModes": ["application/json"],
            "defaultOutputModes": ["application/json"],
            "skills": all_skills,
        }

    def build_app(self, host: str = "localhost", port: int = 8787):
        """
        Build a Starlette app with all agents mounted at their prefixes.

        Args:
            host: Hostname used to build the merged agent card URL.
            port: Port used to build the merged agent card URL.

        Returns:
            A configured Starlette application.
        """
        merged_card = self.build_merged_card(host, port)

        async def merged_card_handler(request: Request) -> JSONResponse:
            return JSONResponse(merged_card)

        routes = [
            Route("/.well-known/agent.json", endpoint=merged_card_handler),
        ]

        for prefix, agent in self._mounts:
            sub_app = agent._build_app(host, port)
            routes.append(Mount(prefix, app=sub_app))

        return Starlette(routes=routes)

    def run(
        self,
        host: str = "0.0.0.0",
        port: int = 8787,
        log_level: str = "info",
    ) -> None:
        """
        Start the router server.

        Args:
            host: Host to bind to.
            port: Port to listen on.
            log_level: Uvicorn log level.
        """
        display_host = "localhost" if host == "0.0.0.0" else host

        print(f"Starting A2A Lite Router on http://{display_host}:{port}")
        for prefix, agent in self._mounts:
            print(f"  {prefix} -> {agent.name} ({len(agent._skills)} skills)")

        app = self.build_app(display_host, port)
        uvicorn.run(app, host=host, port=port, log_level=log_level)
